File: ai-platform/app/services/provider_factory.py
from app.services.provider_interface import ProviderInterface
from app.services.providers import OpenAIProvider, GroqProvider, OllamaProvider, GeminiProvider
from app.core.exceptions import AIPlatformException

class ProviderFactory:
    _providers = {
        "openai": OpenAIProvider,
        "groq": GroqProvider,
        "ollama": OllamaProvider,
        "gemini": GeminiProvider
    }

    @classmethod
    def get_provider(cls, provider_name: str) -> ProviderInterface:
        provider_class = cls._providers.get(provider_name.lower())
        if not provider_class:
            raise AIPlatformException(f"Unsupported AI provider: {provider_name}", status_code=400)
        return provider_class()

File: lms-backend/src/ai/providers/ProviderFactory.js
const PROVIDERS = require('../constants/providers');
const GeminiProvider = require('./GeminiProvider');
const OpenAIProvider = require('./OpenAIProvider');
const GroqProvider = require('./GroqProvider');
const MockProvider = require('./MockProvider');
const aiConfig = require('../config/aiConfig');

/**
 * Provider Factory
 * Returns the appropriate provider instance based on configuration.
 */
class ProviderFactory {
  /**
   * Get an instance of the requested provider.
   * @param {string} [providerName] - Optional provider name. Defaults to the configured default provider.
   * @returns {BaseProvider}
   */
  static getProvider(providerName = aiConfig.defaultProvider) {
    if (!aiConfig.isConfigured) {
      throw new Error('AI Module is not configured or is disabled.');
    }

    switch (providerName) {
      case PROVIDERS.GEMINI:
        return new GeminiProvider(aiConfig);
      case PROVIDERS.OPENAI:
        return new OpenAIProvider(aiConfig);
      case PROVIDERS.GROQ:
        return new GroqProvider(aiConfig);
      case PROVIDERS.MOCK:
        return new MockProvider(aiConfig);
      default:
        throw new Error(`Unsupported AI provider: ${providerName}`);
    }
  }
}

module.exports = ProviderFactory;
